server/src/main/java/org/opensearch/index/shard/RefreshListeners.java

```diff
@@ -256,51 +334,103 @@ public void afterRefresh(boolean didRefresh) throws IOException {
         /* Grab the current refresh listeners and replace them with null while synchronized. Any listeners that come in after this won't be
          * in the list we iterate over and very likely won't be candidates for refresh anyway because we've already moved the
          * lastRefreshedLocation. */
-        List<Tuple<Translog.Location, Consumer<Boolean>>> candidates;
-        synchronized (this) {
-            candidates = refreshListeners;
-            // No listeners to check so just bail early
-            if (candidates == null) {
-                return;
+        if (indexShard != null && indexShard.indexSettings.isSegRepEnabled() && indexShard.routingEntry().primary() == false) {
+            List<Tuple<Long, Consumer<Boolean>>> candidates;
+            synchronized (this) {
+                candidates = seqNoRefreshListeners;
+                // No listeners to check so just bail early
+                if (candidates == null) {
+                    return;
+                }
+                seqNoRefreshListeners = null;
             }
-            refreshListeners = null;
-        }
-        // Iterate the list of listeners, copying the listeners to fire to one list and those to preserve to another list.
-        List<Tuple<Translog.Location, Consumer<Boolean>>> listenersToFire = null;
-        List<Tuple<Translog.Location, Consumer<Boolean>>> preservedListeners = null;
-        for (Tuple<Translog.Location, Consumer<Boolean>> tuple : candidates) {
-            Translog.Location location = tuple.v1();
-            if (location.compareTo(currentRefreshLocation) <= 0) {
-                if (listenersToFire == null) {
-                    listenersToFire = new ArrayList<>();
+            // Iterate the list of listeners, copying the listeners to fire to one list and those to preserve to another list.
+            List<Tuple<Long, Consumer<Boolean>>> listenersToFire = null;
+            List<Tuple<Long, Consumer<Boolean>>> preservedListeners = null;
+
+            for (Tuple<Long, Consumer<Boolean>> tuple : candidates) {
+                Long seqNo = tuple.v1();
+                // logger.info("seqNo is {} and replication checkpoint is {} ",seqNo,
+                // indexShard.getLatestReplicationCheckpoint().getSeqNo());
+                if (seqNo <= indexShard.getLatestReplicationCheckpoint().getSeqNo()) {
```

Comment:
to avoid passing the shard, you can pass IndexShard::getLatestReplicationCheckpoint to this class and invoke as needed.